# Hacker News Analysis
This project will analysis hacker news posts. First step is to open and explore the data.

In [1]:
from csv import reader
hn = list(reader(open('hacker_news.csv')))
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Next, we'll separate the headers from the data.

In [2]:
headers = hn[0]
del hn[0]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Now, we split the posts into three different types, "ask posts", "show posts" and everything else.

In [3]:
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title=row[1]
    if str.startswith(title.lower(),'ask hn'):
        ask_posts.append(row)
    elif str.startswith(title.lower(),'show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
print(len(ask_posts), len(show_posts), len(other_posts))

1744 1162 17194


Let's find out which types of posts receive the most comments on average.

In [7]:
total_ask_comments = 0
for row in ask_posts:
    total_ask_comments += int(row[4])
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)
print(total_ask_comments)
total_show_comments = 0
for row in show_posts:
    total_show_comments += int(row[4])
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)


14.038417431192661
24483
10.31669535283993


On average, ask posts receive more comments (14.0) than show posts (10.3). The analysis will therefore proceed on ask posts only.

Next, we determine the amount of ask posts created at each hour and their total comments, in order to find out at which time they are most popular.

In [9]:
import datetime as dt
result_list = []
for row in ask_posts:
    result_list.append([row[6], int(row[4])])
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    creation_date = dt.datetime.strptime(row[0], '%m/%d/%Y %H:%M')
    hour = creation_date.strftime('%-H')
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]

print(counts_by_hour)
print(comments_by_hour)
print(sum(comments_by_hour.values()))

{'19': 110, '14': 107, '18': 109, '10': 59, '5': 46, '1': 60, '20': 80, '16': 108, '15': 116, '12': 73, '17': 100, '6': 44, '8': 48, '21': 109, '13': 85, '3': 54, '9': 45, '4': 47, '23': 68, '7': 34, '22': 71, '0': 55, '11': 58, '2': 58}
{'19': 1188, '14': 1416, '18': 1439, '10': 793, '5': 464, '1': 683, '20': 1722, '16': 1814, '15': 4477, '12': 687, '17': 1146, '6': 397, '8': 492, '21': 1745, '13': 1253, '3': 421, '9': 251, '4': 337, '23': 543, '7': 267, '22': 479, '0': 447, '11': 641, '2': 1381}
24483


In [11]:
avg_by_hour = []
for hour in counts_by_hour:
    avg = comments_by_hour[hour] / counts_by_hour[hour]
    avg_by_hour.append([hour, avg])
print(avg_by_hour)

[['19', 10.8], ['14', 13.233644859813085], ['18', 13.20183486238532], ['10', 13.440677966101696], ['5', 10.08695652173913], ['1', 11.383333333333333], ['20', 21.525], ['16', 16.796296296296298], ['15', 38.5948275862069], ['12', 9.41095890410959], ['17', 11.46], ['6', 9.022727272727273], ['8', 10.25], ['21', 16.009174311926607], ['13', 14.741176470588234], ['3', 7.796296296296297], ['9', 5.5777777777777775], ['4', 7.170212765957447], ['23', 7.985294117647059], ['7', 7.852941176470588], ['22', 6.746478873239437], ['0', 8.127272727272727], ['11', 11.051724137931034], ['2', 23.810344827586206]]


In [13]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[10.8, '19'], [13.233644859813085, '14'], [13.20183486238532, '18'], [13.440677966101696, '10'], [10.08695652173913, '5'], [11.383333333333333, '1'], [21.525, '20'], [16.796296296296298, '16'], [38.5948275862069, '15'], [9.41095890410959, '12'], [11.46, '17'], [9.022727272727273, '6'], [10.25, '8'], [16.009174311926607, '21'], [14.741176470588234, '13'], [7.796296296296297, '3'], [5.5777777777777775, '9'], [7.170212765957447, '4'], [7.985294117647059, '23'], [7.852941176470588, '7'], [6.746478873239437, '22'], [8.127272727272727, '0'], [11.051724137931034, '11'], [23.810344827586206, '2']]


In [16]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print("Top 5 Hours for Ask Posts Comments")
for avg, hour in sorted_swap[:5]:
    hour = dt.datetime.strptime(hour, '%H')
    hour = hour.strftime('%H:%M')
    print("{}: {:.2f} average comments per post.".format(hour, avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post.
02:00: 23.81 average comments per post.
20:00: 21.52 average comments per post.
16:00: 16.80 average comments per post.
21:00: 16.01 average comments per post.


Posts created at 3pm EST have the most average comments per post.